Replace direct SecDispatcher usage with Maven's SettingsDecrypter API for password decryption

## Description

### Background

- #3859 reported that `k8s:build`/`k8s:push`/`k8s:watch` crash on Maven 3.9.13+ with
  `NoSuchElementException: role: org.sonatype.plexus.components.sec.dispatcher.SecDispatcher, roleHint: maven`,
  because apache/maven#11711 removed the `"maven"`-hinted Plexus XML descriptor from `maven-core`.
- #3860 fixed the crash by replacing the runtime lookup with
  `@Component(role = SecDispatcher.class, hint = "default")` field injection in `AbstractDockerMojo`
  (mirroring the pattern `AbstractJKubeMojo` has used since 2021).

That fix is correct for the crash, but the `"default"`-hinted `DefaultSecDispatcher` is **not** the
dispatcher Maven configures for settings decryption:

| | Removed `"maven"` hint (pre-3.9.13 maven-core) | `"default"` hint (what JKube now injects) |
|---|---|---|
| Master password file | `~/.m2/settings-security.xml` | `~/.settings-security.xml` |
| Matches `mvn --encrypt-master-password` + [Maven Password Encryption guide](https://maven.apache.org/guides/mini/guide-encryption.html) | ✅ | ❌ |
| `-Dsettings.security` override honored | ✅ | ✅ |

The component descriptors confirm this: `plexus-sec-dispatcher-1.4.jar!/META-INF/plexus/components.xml`
declares `<_configuration-file>~/.settings-security.xml</_configuration-file>` for the `default` hint,
while the descriptor removed by apache/maven#11711 declared `~/.m2/settings-security.xml` for the
`maven` hint.

Maven's replacement component (`org.apache.maven.internal.secdispatcher.SecDispatcherProvider`,
since 3.9.13) is explicit about this in its Javadoc:

> Provides a security dispatcher named "maven". Maven uses this security dispatcher instead of the
> default. **The default security dispatcher is unaware of Maven specific configuration file locations
> and should not be used (injected or referenced in any way) at all in any Maven related codebase.**

`HelmPushMojo` already works around this exact problem by force-repointing the injected dispatcher
(`HelmPushMojo.java:38-39` calls `setConfigurationFile(getHelm().getSecurity())`, whose default is
`~/.m2/settings-security.xml`). The docker goals have no such workaround.

docker-maven-plugin hit the identical issue (fabric8io/docker-maven-plugin#1845) and fixed it in
fabric8io/docker-maven-plugin#1848 (authored by @cstamas, the sec-dispatcher maintainer) by switching
to Maven's official **`org.apache.maven.settings.crypto.SettingsDecrypter`** API, which:

- delegates to whatever sec dispatcher the running Maven has correctly configured
  (`~/.m2/settings-security.xml` + `-Dsettings.security` override),
- is a stable Maven API that also works on Maven 4,
- removes the need to depend on `org.sonatype.plexus:plexus-sec-dispatcher` types at all.

### Problem statement

1. **Encrypted registry passwords silently fail to decrypt.** Users who follow JKube's own docs
   (`kubernetes-maven-plugin/doc/src/main/asciidoc/inc/_authentication.adoc`, "Password encryption"
   section) and keep their master password in `~/.m2/settings-security.xml` get a decryption failure,
   after which the **literal `{...}` encrypted blob is sent to the registry as the password**,
   producing a confusing `401 Unauthorized` far away from the root cause.
2. **The error log hides the cause.** `decrypt()` logs the fixed string
   `Failure in decrypting password` without the exception message, the file it looked for, or any hint
   about `-Dsettings.security`.
3. **Duplicated security-sensitive code.** `decrypt()` + the `securityDispatcher` field exist
   byte-for-byte identically in `AbstractDockerMojo` and `AbstractJKubeMojo`.
4. **Workaround code in `HelmPushMojo`** (`instanceof DefaultSecDispatcher` + `setConfigurationFile`)
   exists only to compensate for the wrong default location and can be deleted once `SettingsDecrypter`
   is used.

## Steps to Reproduce

Verified against the `master` build after #3860 (1.20-SNAPSHOT), Maven 3.9.15, macOS:

```bash
# 1. Standard documented setup
mvn --encrypt-master-password reproMaster123     # → {master blob}
cat > ~/.m2/settings-security.xml <<EOF
<settingsSecurity>
  <master>{master blob}</master>
</settingsSecurity>
EOF
ENC=$(mvn --encrypt-password testpass | tail -1)  # → {rZYkML+Q2hcHmdRCuWW1Gh4YfhVcExgX6azrFlghuOM=}

# 2. Any project with an <images> configuration (jib strategy used to avoid needing a Docker daemon)
mvn package org.eclipse.jkube:kubernetes-maven-plugin:1.20-SNAPSHOT:build \
    -Djkube.build.strategy=jib \
    -Djkube.docker.username=testuser \
    "-Djkube.docker.password=$ENC"
```

## Expected Behavior

The password is decrypted using the master password from `~/.m2/settings-security.xml`
(or `-Dsettings.security`), exactly as `mvn --encrypt-password` itself does, and the decrypted value
is used for registry authentication.

## Actual Behavior

```
[ERROR] k8s: Failure in decrypting password
...
[ERROR] ... Failed to authenticate with registry registry-1.docker.io/library/busybox because: 401 Unauthorized
[ERROR] {"details":"incorrect username or password"}
```

The encrypted blob itself is sent to the registry as the password. Adding
`-Dsettings.security=$HOME/.m2/settings-security.xml` makes decryption succeed (confirmed), which is
the interim workaround.

## Proposed Implementation

> Target: `kubernetes-maven-plugin/plugin` module only (OpenShift mojos inherit via `extends`).
> The Gradle plugins do not use SecDispatcher and are out of scope.

### 1. `AbstractDockerMojo.java`

Replace:

```java
@Component(role = org.sonatype.plexus.components.sec.dispatcher.SecDispatcher.class, hint = "default")
protected SecDispatcher securityDispatcher;
```

with:

```java
@Component
protected SettingsDecrypter settingsDecrypter;
```

and replace `decrypt(String)` with:

```java
String decrypt(String password) {
    final Server stub = new Server();
    stub.setPassword(password);
    final SettingsDecryptionResult result =
        settingsDecrypter.decrypt(new DefaultSettingsDecryptionRequest(stub));
    for (SettingsProblem problem : result.getProblems()) {
        getKitLogger().error("Failed to decrypt password: %s", problem);
    }
    return result.getServer().getPassword();
}
```

Notes for the implementer:
- Imports: `org.apache.maven.settings.Server`, `org.apache.maven.settings.building.SettingsProblem`,
  `org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest`,
  `org.apache.maven.settings.crypto.SettingsDecrypter`,
  `org.apache.maven.settings.crypto.SettingsDecryptionResult`.
  Remove the `org.sonatype.plexus.components.sec.dispatcher.*` imports.
- `DefaultSettingsDecrypter` never throws on bad input; it collects `SettingsProblem`s and leaves the
  password as-is, so the existing "return original password on failure" semantics are preserved —
  but now the problem (including the underlying `SecDispatcherException` message) is logged.
- `result.getServer()` is never null when the request contained a server
  (`DefaultSettingsDecryptionResult` returns the first server).

### 2. `AbstractJKubeMojo.java`

Apply the identical replacement to the `securityDispatcher` field (lines 147-148) and `decrypt()`
(lines 262-269). The `passwordDecryptionMethod(this::decrypt)` call sites
(`initJKubeServiceHubBuilder`, `AbstractDockerMojo#getRegistryConfig`) are unchanged.

### 3. `HelmPushMojo.java`

Delete the workaround block in `executeInternal()`:

```java
if (securityDispatcher instanceof DefaultSecDispatcher) {
    ((DefaultSecDispatcher) securityDispatcher).setConfigurationFile(getHelm().getSecurity());
}
```

and the now-unused `org.sonatype.plexus.components.sec.dispatcher.DefaultSecDispatcher` import.

**Decision point:** the `<helm><security>` parameter / `jkube.helm.security` property
(`HelmServiceUtil.java`, default `~/.m2/settings-security.xml`) only exists to feed this workaround.
Recommendation: deprecate it (log a warning when set to a non-default value, point users to Maven's
standard `-Dsettings.security` property) rather than silently ignoring it. Update
`jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc` accordingly.

### 4. Tests

- `AbstractDockerMojoTest` (added in #3860): replace the `SecDispatcher` mock with a
  `SettingsDecrypter` mock. Since the project compiles against `maven-core` 3.8.1 (whose
  `DefaultSettingsDecrypter` is field-injected, no public constructor taking `SecDispatcher`),
  stub the API directly:

  ```java
  settingsDecrypter = mock(SettingsDecrypter.class);
  // success case
  Server decrypted = new Server();
  decrypted.setPassword("decrypted");
  when(settingsDecrypter.decrypt(any())).thenReturn(new DefaultSettingsDecryptionResult(
      Collections.singletonList(decrypted), Collections.emptyList(), Collections.emptyList()));
  // failure case: result containing the original password + a SettingsProblem; assert the
  // original password is returned AND verify(mojo.log).error(...) is called (the spy added in
  // #3860 is currently never verified).
  ```

- `HelmPushMojoTest`: update `helmPushMojo.securityDispatcher = mock(SecDispatcher.class)`
  (line 75) and any assertions tied to the removed workaround.
- Sweep for remaining references: `grep -rn "securityDispatcher\|SecDispatcher" kubernetes-maven-plugin openshift-maven-plugin` must come back empty in `src/main` afterwards.

### 5. Documentation

- Fix the pre-existing filename typo `~/.m2/security-settings.xml` → `~/.m2/settings-security.xml` in:
  - `kubernetes-maven-plugin/doc/src/main/asciidoc/inc/_authentication.adoc` (line 157)
  - `jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm_push.adoc` (line 53)
- Document the deprecation of `jkube.helm.security` if the recommendation in step 3 is accepted.

## Acceptance Criteria

- [ ] No `org.sonatype.plexus.components.sec.dispatcher` imports remain in `src/main` of
      `kubernetes-maven-plugin` / `openshift-maven-plugin`.
- [ ] Encrypted password + master password in `~/.m2/settings-security.xml` decrypts correctly on
      Maven 3.8.x, 3.9.12, and 3.9.15+ (reproducer above; jib strategy needs no Docker daemon).
- [ ] `-Dsettings.security=<file>` override still works.
- [ ] Decryption failure logs the underlying problem/cause, not just a fixed string.
- [ ] `k8s:helm-push` with encrypted repository credentials still works without the
      `setConfigurationFile` workaround.
- [ ] `decrypt()` exists in a single place (or two trivially identical delegations if hierarchy
      merge is deferred — see TODO at `AbstractJKubeMojo.java:240-241`).
- [ ] CHANGELOG entry added.

## Tests

- Unit: `AbstractDockerMojoTest` / `AbstractJKubeMojo`-level tests for success, failure (problem
  logged + original returned), and plaintext pass-through via mocked `SettingsDecrypter`.
- Manual/IT: run the reproducer matrix from "Steps to Reproduce" on Maven 3.9.15 and one
  pre-3.9.13 version; both must decrypt successfully.

## Additional Information

- #3859 (original crash), #3860 (first-step fix)
- apache/maven#11711 (the upstream change; `SecDispatcherProvider` Javadoc quoted above)
- fabric8io/docker-maven-plugin#1845 / fabric8io/docker-maven-plugin#1848 (reference implementation)
- Workaround until fixed: `-Dsettings.security=$HOME/.m2/settings-security.xml`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,7 +22,7 @@ Usage:
 ```
 
 ### 1.20-SNAPSHOT
-* Fix #3859: kubernetes-maven-plugin compatibility with Maven 3.9.13+ SecDispatcher changes
+* Fix #3859: kubernetes-maven-plugin compatibility with Maven 3.9.13+ SecDispatcher changes (decrypt registry/Helm passwords via Maven's SettingsDecrypter; deprecate `<helm><security>`/`jkube.helm.security` in favor of `-Dsettings.security`)
 * Fix #3834: Improve docs for jkube-volume-permission enricher
 * Fix #3848: fix kube-api-test binary download failures due to deprecated kubebuilder-tools storage
 * Fix #2758: Inconsistency when merging fragments with same file name with different profiles

--- a/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
+++ b/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
@@ -173,9 +173,9 @@ Defaults to empty string.
 | `jkube.helm.chartExtension`
 
 ifeval::["{plugin-type}" == "maven"]
-| *security*
-| The Maven security dispatcher configuration file. If you use the default security dispatcher, you need to point this to the file containing your master password. If you followed the http://maven.apache.org/guides/mini/guide-encryption.html[Maven Password Encryption guide], this is `${user.home}/.m2/settings-security.xml`, so that is what this setting defaults to when you do not set it explicitly.
-|
+| *security* _(deprecated)_
+| *Deprecated*: use Maven's `-Dsettings.security=<file>` property instead. This setting will be removed in a future version.
+| `jkube.helm.security`
 endif::[]
 
 | *<<helm-dependencies, dependencies>>*

--- a/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
+++ b/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm.adoc
@@ -175,7 +175,7 @@ Defaults to empty string.
 ifeval::["{plugin-type}" == "maven"]
 | *security* _(deprecated)_
 | *Deprecated*: use Maven's `-Dsettings.security=<file>` property instead. This setting will be removed in a future version.
-| `jkube.helm.security`
+|
 endif::[]
 
 | *<<helm-dependencies, dependencies>>*

--- a/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm_push.adoc
+++ b/jkube-kit/doc/src/main/asciidoc/inc/helm/_jkube_helm_push.adoc
@@ -50,18 +50,9 @@ need to add a server entry for your repo like this:
 </settings>
 ----
 
-If you have encrypted your password with a master password (as outlined in the http://maven.apache.org/guides/mini/guide-encryption.html[Maven Password Encryption guide]), your master password is encrypted somewhere in a file. By default this is `~/.m2/security-settings.xml`. If you've chosen another location, make sure to configure the `security` setting:
-[source,xml,indent=0,subs="verbatim,quotes,attributes"]
-----
-<plugin>
-  <configuration>
-    <helm>
-      <security>~/work/.m2/work-security-settings.xml</security>
-      ...
-    </helm>
-  </configuration>
-</plugin>
-----
+If you have encrypted your password with a master password (as outlined in the http://maven.apache.org/guides/mini/guide-encryption.html[Maven Password Encryption guide]), your master password is encrypted somewhere in a file. By default this is `~/.m2/settings-security.xml`. If you've chosen another location, use Maven's standard `-Dsettings.security=<file>` property to point to it.
+
+NOTE: The `<security>` helm configuration element and `jkube.helm.security` property are deprecated and will be removed in a future version. Use `-Dsettings.security=<file>` instead.
 endif::[]
 
 

--- a/jkube-kit/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmServiceUtil.java
+++ b/jkube-kit/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmServiceUtil.java
@@ -75,7 +75,7 @@ public class HelmServiceUtil {
   protected static final String STABLE_REPOSITORY = "stableRepository";
   protected static final String SNAPSHOT_REPOSITORY = "snapshotRepository";
   protected static final String PROPERTY_SECURITY =  "jkube.helm.security";
-  protected static final String DEFAULT_SECURITY = "~/.m2/settings-security.xml";
+  public static final String DEFAULT_SECURITY = "~/.m2/settings-security.xml";
 
   protected static final String PROPERTY_HELM_LINT_STRICT = "jkube.helm.lint.strict";
   protected static final String PROPERTY_HELM_LINT_QUIET = "jkube.helm.lint.quiet";

--- a/kubernetes-maven-plugin/doc/src/main/asciidoc/inc/_authentication.adoc
+++ b/kubernetes-maven-plugin/doc/src/main/asciidoc/inc/_authentication.adoc
@@ -154,7 +154,7 @@ Regardless of which mode you choose you can encrypt password as described
 in the
 http://maven.apache.org/guides/mini/guide-encryption.html[Maven documentation]. Assuming
 that you have setup a _master password_ in
-`~/.m2/security-settings.xml` you can create easily encrypt
+`~/.m2/settings-security.xml` you can create easily encrypt
 passwords:
 
 .Example

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
@@ -72,6 +72,7 @@ import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
 import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.fusesource.jansi.Ansi;
+import javax.inject.Inject;
 
 import static org.eclipse.jkube.kit.build.service.docker.DockerAccessFactory.DockerAccessContext.DEFAULT_MAX_CONNECTIONS;
 import static org.eclipse.jkube.kit.common.util.BuildReferenceDateUtil.getBuildTimestamp;
@@ -316,8 +317,7 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     // Mode which is resolved, also when 'auto' is set
     protected RuntimeMode runtimeMode;
 
-    @SuppressWarnings("deprecation")
-    @Component
+    @Inject
     protected SettingsDecrypter settingsDecrypter;
 
     /**

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
@@ -66,9 +66,12 @@ import org.apache.maven.project.MavenProjectHelper;
 import org.apache.maven.settings.Settings;
 import org.apache.maven.shared.utils.logging.MessageUtils;
 import org.eclipse.jkube.maven.plugin.mojo.KitLoggerProvider;
+import org.apache.maven.settings.Server;
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.fusesource.jansi.Ansi;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcher;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcherException;
 
 import static org.eclipse.jkube.kit.build.service.docker.DockerAccessFactory.DockerAccessContext.DEFAULT_MAX_CONNECTIONS;
 import static org.eclipse.jkube.kit.common.util.BuildReferenceDateUtil.getBuildTimestamp;
@@ -313,8 +316,8 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     // Mode which is resolved, also when 'auto' is set
     protected RuntimeMode runtimeMode;
 
-    @Component(role = org.sonatype.plexus.components.sec.dispatcher.SecDispatcher.class, hint = "default")
-    protected SecDispatcher securityDispatcher;
+    @Component
+    protected SettingsDecrypter settingsDecrypter;
 
     /**
      * Watching mode for rebuilding images
@@ -438,12 +441,15 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     }
 
     String decrypt(String password) {
-        try {
-            return securityDispatcher.decrypt(password);
-        } catch (SecDispatcherException e) {
-            getKitLogger().error("Failure in decrypting password");
+        final Server stub = new Server();
+        stub.setPassword(password);
+        final SettingsDecryptionResult result =
+            settingsDecrypter.decrypt(new DefaultSettingsDecryptionRequest(stub));
+        for (SettingsProblem problem : result.getProblems()) {
+            getKitLogger().error("Failed to decrypt password: %s", problem);
         }
-        return password;
+        final Server decrypted = result.getServer();
+        return decrypted != null ? decrypted.getPassword() : password;
     }
 
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
@@ -317,9 +317,12 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     // Mode which is resolved, also when 'auto' is set
     protected RuntimeMode runtimeMode;
 
-    @SuppressWarnings("java:S6813") // Maven Mojos don't support constructor injection
-    @Inject
     protected SettingsDecrypter settingsDecrypter;
+
+    @Inject
+    void setSettingsDecrypter(SettingsDecrypter settingsDecrypter) {
+        this.settingsDecrypter = settingsDecrypter;
+    }
 
     /**
      * Watching mode for rebuilding images

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
@@ -317,6 +317,7 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     // Mode which is resolved, also when 'auto' is set
     protected RuntimeMode runtimeMode;
 
+    @SuppressWarnings("java:S6813") // Maven Mojos don't support constructor injection
     @Inject
     protected SettingsDecrypter settingsDecrypter;
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java
@@ -316,6 +316,7 @@ public abstract class AbstractDockerMojo extends AbstractMojo
     // Mode which is resolved, also when 'auto' is set
     protected RuntimeMode runtimeMode;
 
+    @SuppressWarnings("deprecation")
     @Component
     protected SettingsDecrypter settingsDecrypter;
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
@@ -48,8 +48,11 @@ import org.eclipse.jkube.kit.config.resource.RuntimeMode;
 import org.eclipse.jkube.kit.config.service.JKubeServiceHub;
 import org.eclipse.jkube.kit.resource.service.DefaultResourceService;
 import org.eclipse.jkube.maven.plugin.mojo.KitLoggerProvider;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcher;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcherException;
+import org.apache.maven.settings.Server;
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 
 import java.io.File;
 import java.io.IOException;
@@ -144,8 +147,8 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     @Parameter
     protected ClusterConfiguration access;
 
-    @Component(role = org.sonatype.plexus.components.sec.dispatcher.SecDispatcher.class, hint = "default")
-    protected SecDispatcher securityDispatcher;
+    @Component
+    protected SettingsDecrypter settingsDecrypter;
 
     protected KitLogger log;
 
@@ -260,12 +263,15 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     }
 
     String decrypt(String password) {
-        try {
-            return securityDispatcher.decrypt(password);
-        } catch (SecDispatcherException e) {
-            getKitLogger().error("Failure in decrypting password");
+        final Server stub = new Server();
+        stub.setPassword(password);
+        final SettingsDecryptionResult result =
+            settingsDecrypter.decrypt(new DefaultSettingsDecryptionRequest(stub));
+        for (SettingsProblem problem : result.getProblems()) {
+            getKitLogger().error("Failed to decrypt password: %s", problem);
         }
-        return password;
+        final Server decrypted = result.getServer();
+        return decrypted != null ? decrypted.getPassword() : password;
     }
 
     protected void cleanWorkDirectory() throws IOException {

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
@@ -147,6 +147,7 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     @Parameter
     protected ClusterConfiguration access;
 
+    @SuppressWarnings("deprecation")
     @Component
     protected SettingsDecrypter settingsDecrypter;
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
@@ -53,6 +53,7 @@ import org.apache.maven.settings.building.SettingsProblem;
 import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
 import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecryptionResult;
+import javax.inject.Inject;
 
 import java.io.File;
 import java.io.IOException;
@@ -147,8 +148,7 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     @Parameter
     protected ClusterConfiguration access;
 
-    @SuppressWarnings("deprecation")
-    @Component
+    @Inject
     protected SettingsDecrypter settingsDecrypter;
 
     protected KitLogger log;

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
@@ -148,6 +148,7 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     @Parameter
     protected ClusterConfiguration access;
 
+    @SuppressWarnings("java:S6813") // Maven Mojos don't support constructor injection
     @Inject
     protected SettingsDecrypter settingsDecrypter;
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojo.java
@@ -148,9 +148,12 @@ public abstract class AbstractJKubeMojo extends AbstractMojo implements KitLogge
     @Parameter
     protected ClusterConfiguration access;
 
-    @SuppressWarnings("java:S6813") // Maven Mojos don't support constructor injection
-    @Inject
     protected SettingsDecrypter settingsDecrypter;
+
+    @Inject
+    void setSettingsDecrypter(SettingsDecrypter settingsDecrypter) {
+        this.settingsDecrypter = settingsDecrypter;
+    }
 
     protected KitLogger log;
 

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java
@@ -18,26 +18,29 @@ import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.ResolutionScope;
-import org.sonatype.plexus.components.sec.dispatcher.DefaultSecDispatcher;
 
 import static org.eclipse.jkube.kit.resource.helm.HelmServiceUtil.initHelmPushConfig;
 
 @Mojo(name = "helm-push", defaultPhase = LifecyclePhase.INSTALL, requiresDependencyResolution = ResolutionScope.COMPILE)
 public class HelmPushMojo extends AbstractHelmMojo {
 
+  private static final String DEFAULT_SECURITY = "~/.m2/settings-security.xml";
+
   @Override
   public void init() throws MojoFailureException {
     super.init();
 
     initHelmPushConfig(helm, javaProject);
+    if (helm.getSecurity() != null && !DEFAULT_SECURITY.equals(helm.getSecurity())) {
+      getKitLogger().warn("The <security> helm configuration and jkube.helm.security property are deprecated" +
+          " and will be removed in a future version." +
+          " Use Maven's -Dsettings.security=<file> property instead.");
+    }
   }
 
   @Override
   public void executeInternal() throws MojoExecutionException {
     try {
-      if (securityDispatcher instanceof DefaultSecDispatcher) {
-        ((DefaultSecDispatcher) securityDispatcher).setConfigurationFile(getHelm().getSecurity());
-      }
       jkubeServiceHub.getHelmService().uploadHelmChart(getHelm());
     } catch (Exception exp) {
       getKitLogger().error("Error performing Helm push", exp);

--- a/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java
+++ b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java
@@ -19,12 +19,11 @@ import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.ResolutionScope;
 
+import static org.eclipse.jkube.kit.resource.helm.HelmServiceUtil.DEFAULT_SECURITY;
 import static org.eclipse.jkube.kit.resource.helm.HelmServiceUtil.initHelmPushConfig;
 
 @Mojo(name = "helm-push", defaultPhase = LifecyclePhase.INSTALL, requiresDependencyResolution = ResolutionScope.COMPILE)
 public class HelmPushMojo extends AbstractHelmMojo {
-
-  private static final String DEFAULT_SECURITY = "~/.m2/settings-security.xml";
 
   @Override
   public void init() throws MojoFailureException {

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojoTest.java
@@ -14,39 +14,49 @@
 package org.eclipse.jkube.maven.plugin.mojo.build;
 
 import java.io.IOException;
+import java.util.Collections;
 
+import org.apache.maven.settings.Server;
 import org.apache.maven.settings.Settings;
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.eclipse.jkube.kit.common.KitLogger;
 import org.eclipse.jkube.kit.common.RegistryConfig;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcher;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcherException;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 class AbstractDockerMojoTest {
 
   private TestMojo mojo;
-  private SecDispatcher secDispatcher;
+  private SettingsDecrypter settingsDecrypter;
 
   @BeforeEach
   void setUp() {
-    secDispatcher = mock(SecDispatcher.class);
+    settingsDecrypter = mock(SettingsDecrypter.class);
     mojo = new TestMojo();
     mojo.settings = new Settings();
-    mojo.securityDispatcher = secDispatcher;
+    mojo.settingsDecrypter = settingsDecrypter;
     mojo.log = spy(new KitLogger.SilentLogger());
   }
 
   @Test
-  void getRegistryConfig_shouldDecryptPasswordUsingSecDispatcher() throws SecDispatcherException {
+  void getRegistryConfig_shouldDecryptPasswordUsingSettingsDecrypter() {
     // Given
-    when(secDispatcher.decrypt("encrypted")).thenReturn("decrypted");
+    Server decrypted = new Server();
+    decrypted.setPassword("decrypted");
+    SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+    when(result.getServer()).thenReturn(decrypted);
+    when(result.getProblems()).thenReturn(Collections.emptyList());
+    when(settingsDecrypter.decrypt(any())).thenReturn(result);
     // When
     RegistryConfig config = mojo.getRegistryConfig(null);
     // Then
@@ -54,13 +64,34 @@ class AbstractDockerMojoTest {
   }
 
   @Test
-  void getRegistryConfig_whenDecryptionFails_shouldReturnOriginalPassword() throws SecDispatcherException {
+  void getRegistryConfig_whenDecryptionFails_shouldReturnOriginalPasswordAndLogProblem() {
     // Given
-    when(secDispatcher.decrypt("encrypted")).thenThrow(new SecDispatcherException("test"));
+    Server original = new Server();
+    original.setPassword("encrypted");
+    SettingsProblem problem = mock(SettingsProblem.class);
+    when(problem.toString()).thenReturn("Failed to decrypt password: test error");
+    SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+    when(result.getServer()).thenReturn(original);
+    when(result.getProblems()).thenReturn(Collections.singletonList(problem));
+    when(settingsDecrypter.decrypt(any())).thenReturn(result);
     // When
     RegistryConfig config = mojo.getRegistryConfig(null);
     // Then
     assertThat(config.getPasswordDecryptionMethod().apply("encrypted")).isEqualTo("encrypted");
+    verify(mojo.log).error("Failed to decrypt password: %s", problem);
+  }
+
+  @Test
+  void getRegistryConfig_whenDecryptionReturnsNullServer_shouldReturnOriginalPassword() {
+    // Given
+    SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+    when(result.getServer()).thenReturn(null);
+    when(result.getProblems()).thenReturn(Collections.emptyList());
+    when(settingsDecrypter.decrypt(any())).thenReturn(result);
+    // When
+    RegistryConfig config = mojo.getRegistryConfig(null);
+    // Then
+    assertThat(config.getPasswordDecryptionMethod().apply("original")).isEqualTo("original");
   }
 
   @Test

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojoTest.java
@@ -69,7 +69,6 @@ class AbstractDockerMojoTest {
     Server original = new Server();
     original.setPassword("encrypted");
     SettingsProblem problem = mock(SettingsProblem.class);
-    when(problem.toString()).thenReturn("Failed to decrypt password: test error");
     SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
     when(result.getServer()).thenReturn(original);
     when(result.getProblems()).thenReturn(Collections.singletonList(problem));

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
@@ -17,6 +17,7 @@ import java.util.Collections;
 
 import org.apache.maven.settings.Server;
 import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
 import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.eclipse.jkube.kit.common.KitLogger;
@@ -25,6 +26,7 @@ import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
@@ -51,7 +53,7 @@ class AbstractJKubeMojoTest {
   class Decrypt {
 
     @Test
-    @DisplayName("with successful decryption, should return decrypted password")
+    @DisplayName("with successful decryption, should return decrypted password and pass input to request")
     void withSuccessfulDecryption() {
       // Given
       Server decrypted = new Server();
@@ -64,6 +66,12 @@ class AbstractJKubeMojoTest {
       String password = mojo.decrypt("encrypted");
       // Then
       assertThat(password).isEqualTo("decrypted");
+      ArgumentCaptor<DefaultSettingsDecryptionRequest> captor =
+          ArgumentCaptor.forClass(DefaultSettingsDecryptionRequest.class);
+      verify(settingsDecrypter).decrypt(captor.capture());
+      assertThat(captor.getValue().getServers())
+          .singleElement()
+          .hasFieldOrPropertyWithValue("password", "encrypted");
     }
 
     @Test
@@ -73,7 +81,6 @@ class AbstractJKubeMojoTest {
       Server original = new Server();
       original.setPassword("encrypted");
       SettingsProblem problem = mock(SettingsProblem.class);
-      when(problem.toString()).thenReturn("Failed to decrypt password: test error");
       SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
       when(result.getServer()).thenReturn(original);
       when(result.getProblems()).thenReturn(Collections.singletonList(problem));

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
@@ -1,0 +1,108 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.maven.plugin.mojo.build;
+
+import java.util.Collections;
+
+import org.apache.maven.settings.Server;
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
+import org.eclipse.jkube.kit.common.KitLogger;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class AbstractJKubeMojoTest {
+
+  private TestMojo mojo;
+  private SettingsDecrypter settingsDecrypter;
+
+  @BeforeEach
+  void setUp() {
+    settingsDecrypter = mock(SettingsDecrypter.class);
+    mojo = new TestMojo();
+    mojo.settingsDecrypter = settingsDecrypter;
+    mojo.log = spy(new KitLogger.SilentLogger());
+  }
+
+  @Nested
+  @DisplayName("decrypt")
+  class Decrypt {
+
+    @Test
+    @DisplayName("with successful decryption, should return decrypted password")
+    void withSuccessfulDecryption() {
+      // Given
+      Server decrypted = new Server();
+      decrypted.setPassword("decrypted");
+      SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+      when(result.getServer()).thenReturn(decrypted);
+      when(result.getProblems()).thenReturn(Collections.emptyList());
+      when(settingsDecrypter.decrypt(any())).thenReturn(result);
+      // When
+      String password = mojo.decrypt("encrypted");
+      // Then
+      assertThat(password).isEqualTo("decrypted");
+    }
+
+    @Test
+    @DisplayName("with decryption failure, should return original password and log problem")
+    void withDecryptionFailure() {
+      // Given
+      Server original = new Server();
+      original.setPassword("encrypted");
+      SettingsProblem problem = mock(SettingsProblem.class);
+      when(problem.toString()).thenReturn("Failed to decrypt password: test error");
+      SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+      when(result.getServer()).thenReturn(original);
+      when(result.getProblems()).thenReturn(Collections.singletonList(problem));
+      when(settingsDecrypter.decrypt(any())).thenReturn(result);
+      // When
+      String password = mojo.decrypt("encrypted");
+      // Then
+      assertThat(password).isEqualTo("encrypted");
+      verify(mojo.log).error("Failed to decrypt password: %s", problem);
+    }
+
+    @Test
+    @DisplayName("with null server result, should return original password")
+    void withNullServerResult() {
+      // Given
+      SettingsDecryptionResult result = mock(SettingsDecryptionResult.class);
+      when(result.getServer()).thenReturn(null);
+      when(result.getProblems()).thenReturn(Collections.emptyList());
+      when(settingsDecrypter.decrypt(any())).thenReturn(result);
+      // When
+      String password = mojo.decrypt("original");
+      // Then
+      assertThat(password).isEqualTo("original");
+    }
+  }
+
+  private static class TestMojo extends AbstractJKubeMojo {
+    @Override
+    public void executeInternal() {
+    }
+  }
+}

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
@@ -13,19 +13,27 @@
  */
 package org.eclipse.jkube.maven.plugin.mojo.build;
 
+import java.lang.reflect.Field;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.Collections;
 
 import org.apache.maven.settings.Server;
 import org.apache.maven.settings.building.SettingsProblem;
 import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.DefaultSettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.eclipse.jkube.kit.common.KitLogger;
+import org.sonatype.plexus.components.cipher.DefaultPlexusCipher;
+import org.sonatype.plexus.components.sec.dispatcher.DefaultSecDispatcher;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.ArgumentCaptor;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -104,6 +112,29 @@ class AbstractJKubeMojoTest {
       String password = mojo.decrypt("original");
       // Then
       assertThat(password).isEqualTo("original");
+    }
+
+    @Test
+    @DisplayName("with a real SettingsDecrypter, decrypts an encrypted password")
+    void decrypt_withRealDecrypter_returnsPlaintext(@TempDir Path tempDir) throws Exception {
+      // Given
+      final DefaultPlexusCipher cipher = new DefaultPlexusCipher();
+      final String master = "masterPassword";
+      final Path securityFile = tempDir.resolve("settings-security.xml");
+      Files.write(securityFile, ("<settingsSecurity><master>"
+          + cipher.encryptAndDecorate(master, DefaultSecDispatcher.SYSTEM_PROPERTY_SEC_LOCATION)
+          + "</master></settingsSecurity>").getBytes(StandardCharsets.UTF_8));
+      final String encryptedPassword = cipher.encryptAndDecorate("s3cr3t", master);
+      final DefaultSecDispatcher secDispatcher = new DefaultSecDispatcher();
+      final Field cipherField = DefaultSecDispatcher.class.getDeclaredField("_cipher");
+      cipherField.setAccessible(true);
+      cipherField.set(secDispatcher, cipher);
+      secDispatcher.setConfigurationFile(securityFile.toAbsolutePath().toString());
+      mojo.settingsDecrypter = new DefaultSettingsDecrypter(secDispatcher);
+      // When
+      final String decrypted = mojo.decrypt(encryptedPassword);
+      // Then
+      assertThat(decrypted).isEqualTo("s3cr3t");
     }
   }
 

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractJKubeMojoTest.java
@@ -103,6 +103,7 @@ class AbstractJKubeMojoTest {
   private static class TestMojo extends AbstractJKubeMojo {
     @Override
     public void executeInternal() {
+      // Not needed for unit testing decrypt()
     }
   }
 }

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojoTest.java
@@ -14,10 +14,14 @@
 package org.eclipse.jkube.maven.plugin.mojo.build;
 
 import java.nio.file.Path;
+import java.util.Collections;
 import java.util.HashMap;
 
 import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugin.descriptor.PluginDescriptor;
+import org.apache.maven.settings.Server;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 import org.eclipse.jkube.kit.common.KitLogger;
 import org.eclipse.jkube.kit.common.RegistryServerConfiguration;
 import org.eclipse.jkube.kit.resource.helm.BadUploadException;
@@ -29,21 +33,17 @@ import org.apache.maven.plugin.MojoExecution;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.descriptor.MojoDescriptor;
 import org.apache.maven.project.MavenProject;
-import org.apache.maven.settings.Server;
 import org.apache.maven.settings.Settings;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 import org.mockito.MockedConstruction;
-import org.sonatype.plexus.components.sec.dispatcher.SecDispatcher;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
-import org.mockito.AdditionalAnswers;
 
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.mockConstruction;
@@ -72,7 +72,7 @@ class HelmPushMojoTest {
     helmPushMojo.helm = new HelmConfig();
     helmPushMojo.project = new MavenProject();
     helmPushMojo.settings = new Settings();
-    helmPushMojo.securityDispatcher = mock(SecDispatcher.class);
+    helmPushMojo.settingsDecrypter = mock(SettingsDecrypter.class);
     helmPushMojo.mojoExecution = new MojoExecution(new MojoDescriptor());
     helmPushMojo.interpolateTemplateParameters = true;
     helmPushMojo.project.getBuild()
@@ -82,8 +82,15 @@ class HelmPushMojoTest {
     helmPushMojo.mojoExecution.getMojoDescriptor().setPluginDescriptor(new PluginDescriptor());
     helmPushMojo.mojoExecution.getMojoDescriptor().getPluginDescriptor().setGoalPrefix("k8s");
     helmPushMojo.mojoExecution.getMojoDescriptor().setGoal("helm-push");
-    when(helmPushMojo.securityDispatcher.decrypt(anyString()))
-      .thenReturn(String.valueOf(AdditionalAnswers.returnsFirstArg()));
+    SettingsDecryptionResult decryptionResult = mock(SettingsDecryptionResult.class);
+    Server passthroughServer = new Server();
+    when(decryptionResult.getServer()).thenAnswer(invocation -> {
+      Server s = new Server();
+      s.setPassword("passthrough");
+      return s;
+    });
+    when(decryptionResult.getProblems()).thenReturn(Collections.emptyList());
+    when(helmPushMojo.settingsDecrypter.decrypt(any())).thenReturn(decryptionResult);
   }
 
   @AfterEach
@@ -220,6 +227,22 @@ class HelmPushMojoTest {
       // Then
       assertThat(helmServiceMockedConstruction.constructed()).isEmpty();
       verify(logger, times(1)).info("`%s` goal is skipped.", "k8s:helm-push");
+    }
+  }
+
+  @Test
+  void init_withCustomSecurityConfig_shouldLogDeprecationWarning() throws Exception {
+    try (MockedConstruction<HelmService> helmServiceMockedConstruction = mockConstruction(HelmService.class)) {
+      // Given
+      helmPushMojo.project.getProperties().put("jkube.helm.security", "~/custom/settings-security.xml");
+      helmPushMojo.helm.setSnapshotRepository(completeValidRepository());
+      helmPushMojo.project.setVersion("1337-SNAPSHOT");
+      // When
+      helmPushMojo.execute();
+      // Then
+      verify(logger).warn("The <security> helm configuration and jkube.helm.security property are deprecated" +
+          " and will be removed in a future version." +
+          " Use Maven's -Dsettings.security=<file> property instead.");
     }
   }
 

--- a/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojoTest.java
+++ b/kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojoTest.java
@@ -60,7 +60,7 @@ class HelmPushMojoTest {
   private KitLogger logger;
 
   @BeforeEach
-  void setUp() throws Exception {
+  void setUp() {
     logger = spy(new KitLogger.SilentLogger());
     helmPushMojo = new HelmPushMojo() {
       @Override
@@ -83,7 +83,6 @@ class HelmPushMojoTest {
     helmPushMojo.mojoExecution.getMojoDescriptor().getPluginDescriptor().setGoalPrefix("k8s");
     helmPushMojo.mojoExecution.getMojoDescriptor().setGoal("helm-push");
     SettingsDecryptionResult decryptionResult = mock(SettingsDecryptionResult.class);
-    Server passthroughServer = new Server();
     when(decryptionResult.getServer()).thenAnswer(invocation -> {
       Server s = new Server();
       s.setPassword("passthrough");
@@ -241,6 +240,21 @@ class HelmPushMojoTest {
       helmPushMojo.execute();
       // Then
       verify(logger).warn("The <security> helm configuration and jkube.helm.security property are deprecated" +
+          " and will be removed in a future version." +
+          " Use Maven's -Dsettings.security=<file> property instead.");
+    }
+  }
+
+  @Test
+  void init_withDefaultSecurityConfig_shouldNotLogDeprecationWarning() throws Exception {
+    try (MockedConstruction<HelmService> helmServiceMockedConstruction = mockConstruction(HelmService.class)) {
+      // Given
+      helmPushMojo.helm.setSnapshotRepository(completeValidRepository());
+      helmPushMojo.project.setVersion("1337-SNAPSHOT");
+      // When
+      helmPushMojo.execute();
+      // Then
+      verify(logger, times(0)).warn("The <security> helm configuration and jkube.helm.security property are deprecated" +
           " and will be removed in a future version." +
           " Use Maven's -Dsettings.security=<file> property instead.");
     }

